consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -249,23 +262,44 @@ where
         if let Some(old_qc) = self.id_to_quorum_cert.get(&block_id) {
             return VoteReceptionResult::OldQuorumCertificate(Arc::clone(old_qc));
         }
+
+        let author = vote_msg.author();
+        if let Some(last_voted_block) = self.author_to_last_voted_block_id.get(&author) {
+            if block_id == last_voted_block.block_id {
+                // Author has already voted for this block
+                return VoteReceptionResult::DuplicateVote;
+            }
+            if let Some(block_pending_votes) = self.id_to_votes.get_mut(&last_voted_block.block_id)
+            {
+                if let Some(li_digest_to_sig) =
+                    block_pending_votes.get_mut(&last_voted_block.digest)
+                {
+                    // Removing signature from last voted block
+                    li_digest_to_sig.remove_signature(author);
+                    if li_digest_to_sig.signatures().is_empty() {
+                        // Last vote/signature for block, remove digest entry
+                        block_pending_votes.remove(&last_voted_block.digest);
```

Comment:
Should we also remove the `block_id` entry from `id_to_votes` if `block_pending_votes.is_empty()`?
Otherwise we'd be still vulnerable to an attack that keeps multiple entries with empty hashmaps in `id_to_votes`.

Reply:
Good catch. Fixed in latest commit.